src/go/k8s/internal/controller/redpanda/cluster_controller.go

```diff
@@ -1023,9 +985,43 @@ func (r *ClusterReconciler) doDecommissionGhostBrokers(c context.Context, vClust
 	if err != nil {
 		return fmt.Errorf("unable to create admin client: %w", err)
 	}
-	adminBrokers, err := adminClient.Brokers(ctx)
+
+	health, err := adminClient.GetHealthOverview(ctx)
+	if err != nil {
+		return fmt.Errorf("feating cluster heatlh overview: %w", err)
+	}
+
+	if !health.IsHealthy {
+		err = errors.New("cluster is not healthy")
+		log.V(logger.DebugLevel).Error(err, "stopping decommission verification",
+			"requested-replicas", vCluster.Spec.Replicas,
+			"nodes-registered", health.AllNodes,
+			"nodes-down", health.NodesDown)
+		return err
+	}
+
+	ss, err := ar.getStatefulSet()
 	if err != nil {
-		return fmt.Errorf("unable to fetch brokers: %w", err)
+		return err
+	}
+
+	sts := &appsv1.StatefulSet{}
+	if err := r.Client.Get(ctx, ss.Key(), sts); err != nil {
+		return fmt.Errorf("could not retrieve the statefulset: %w", err)
+	}
+
+	if vCluster.Spec.Replicas == nil {
+		return nil
+	}
+
+	if len(health.AllNodes) <= int(*vCluster.Spec.Replicas) {
```

Comment:
Is `AllNodes` literally all brokers or is it just healthy ones? Either way, I think this check might be incorrect. Wouldn't we want the gating to be "If the total number of healthy Brokers from the admin API is equal to the total number of healthy Pods from the k8s API, it's safe to check for ghost brokers?"

I think we'll benefit from some more comments in here about the intent of the logic. Without that we'll only be able to talk about what the code does not what it was intended to do.

Reply:
> Is AllNodes literally all brokers or is it just healthy ones?

All nodes that are not decommissioned (healthy and unhealthy).

> Wouldn't we want the gating to be "If the total number of healthy Brokers from the admin API is equal to the total number of healthy Pods from the k8s API, it's safe to check for ghost brokers?"

That would work too. Cluster health Redpanda endpoint would not be correct. The broker GET REST endpoint would be better with the additional mapping between Redpanda Node ID and Pod ordinal. 

```
$ rpk redpanda admin brokers list -X admin.tls.enabled=true -X admin.tls.ca=/etc/tls/certs/default/ca.crt -X admin.tls.cert=/etc/tls/certs/default/tls.crt -X admin.tls.key=/etc/tls/certs/default/tls.key
NODE-ID  NUM-CORES  MEMBERSHIP-STATUS  IS-ALIVE  BROKER-VERSION
0        1          active             true      v24.1.1 - b5ade3f40320514e855bacfad3a75d76e25a7326
1        1          active             true      v24.1.1 - b5ade3f40320514e855bacfad3a75d76e25a7326
2        1          active             true      v24.1.1 - b5ade3f40320514e855bacfad3a75d76e25a7326
```

https://github.com/redpanda-data/redpanda/blob/03ccde6ca4d211724d513985669504479a95332f/src/go/rpk/pkg/adminapi/api_broker.go#L45-L52

> I think we'll benefit from some more comments in here about the intent of the logic. Without that we'll only be able to talk about what the code does not what it was intended to do.

Ok, I will change the implementation to use brokers endpoint with additional comments to express intend.